### PR TITLE
Return bytes size on write

### PR DESCRIPTION
**Pull request by stevendikowitz**
Python3 `write()` is expecting to return [number of bytes written](https://docs.python.org/3/library/io.html#io.RawIOBase.write)

### CODE REVIEW

File: smart_open/smart_open_lib.py [hunk 1]
```diff
@@ -833,12 +833,17 @@ def write(self, b):
...
+            buff_len = len(buff)
+            logger.info("uploading part #%i, %i bytes (total %.3fGB)" % (self.parts, buff_len, self.total_size / 1024.0 ** 3))
             self.mp.upload_part_from_file(BytesIO(buff), part_num=self.parts + 1)
             logger.debug("upload of part #%i finished" % self.parts)
             self.parts += 1
             self.lines, self.chunk_bytes = [], 0
 
+            return buff_len
+
+        return 0
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: Is this correct? What is the actual contract for this number -- the number of bytes accepted (then this shouldn't be 0), or the number of bytes passed to lower I/O layer (then 0 OK), or number of bytes actually written/ACK'd/synced to some physical medium...?

File: smart_open/smart_open_lib.py [hunk 2]
```diff
@@ -935,12 +940,17 @@ def write(self, b):
...
+            buff_len = len(buff)
+            logger.info("uploading part #%i, %i bytes (total %.3fGB)" % (self.parts, buff_len, self.total_size / 1024.0 ** 3))
             self.upload(buff)
             logger.debug("upload of part #%i finished" % self.parts)
             self.parts += 1
             self.lines, self.chunk_bytes = [], 0
 
+            return buff_len
+
+        return 0
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: Dtto.

**Comment by piskvorky** [comment]
Thanks @stevendikowitz ! Returning the number of bytes is desirable (and fully backward compatible 👍 ). But your link wasn't clear about the semantics of this number -- can you clarify? See my PR comments.

**Comment by stevendikowitz** [comment]
Hey @piskvorky, thanks for checking it out so soon. Your comment was on point, write() actually needs to be returning `len(b)` because ZipFile is using that length as its offset. I've updated this PR and included #114 into it because I also needed `flush()`. I left a comment on 114 because of how flush was implemented in S3OpenWrite, but it makes sense based on how it was implemented.

**Comment by stevendikowitz** [comment]
Hey @piskvorky, any update on the status of this PR? I'd like to help get this merged down

**Comment by stevendikowitz** [comment]
Closing PR for inaction, please reach out to pick it up again if interested.

**Comment by piskvorky** [comment]
Sorry @stevendikowitz -- @menshikh-iv is a better person to CC on open-source related matters.

@menshikh-iv can you have a look please?

File: smart_open/tests/test_smart_open.py [hunk 3]
```diff
@@ -393,6 +393,18 @@ def test_buffer_flushed_after_eof(self):
 class S3OpenReadTest(unittest.TestCase):
 
     @mock_s3
+    def test_open_and_close(self):
```
Comment (menshikh-iv) [inline smart_open/tests/test_smart_open.py]: Need to add test with asserts for checking return value after `write` call

**Comment by menshikh-iv** [comment]
Sorry for waiting @stevendikowitz, please merge the current master branch to your PR and add a test for checking the return value, that's enough.

**Comment by menshikh-iv** [comment]
ping @stevendikowitz, are you planning to finish this PR?

**Comment by stevendikowitz** [comment]
Hey guys, sorry for delay, real life intrudes. I would like to finish, but I'll be traveling and working on a project this week, I won't have time to head back till the following weekend. Feel free to pick it up, otherwise I'll do my best to hop back on next weekend.

**Comment by menshikh-iv** [comment]
Ping @stevendikowitz, are you planning to finish PR?

**Comment by menshikh-iv** [comment]
ping @stevendikowitz, are you planning to finish PR?

**Comment by menshikh-iv** [comment]
I close PR because it is abandoned

**Pull request closed without merging**